### ISSUE TITLE
Ligand adjacency is off by one atom (predates the PE work, and the PE sits on top of it)

### ISSUE THREAD

**Issue by kuanlinhuang** [Huang-lab/TRACE-Kin#3]
Reviewed `b3682eb` ("added Positional Enconding for ligand graph via RWSE and LPSE") as part of the 2026-08-20 lab code review. The encoder itself is the best-documented module in the repo — the docstring argues for RWSE over Laplacian PE on the right grounds (sign ambiguity, near-degenerate eigenvalues, return probabilities encoding ring size), names the DGL-vs-PyG deviation explicitly, and says what was *not* carried over from the reference. That is unusually careful.

But the encoding is being computed on a ligand graph whose bonds are attached to the wrong atoms, and that predates this commit.

---

## 1. `bond_feature()` shifts every atom index by one

`training/ligand_init.py:283`:

```python
for b in mol.GetBonds():
    v1 = b.GetBeginAtomIdx()
    v2 = b.GetEndAtomIdx()
    b_type = self.edge_dict[b.GetBondType()]
    adj[v1 - 1, v2 - 1] = b_type
    adj[v2 - 1, v1 - 1] = b_type
```

RDKit atom indices are **0-based**. `- 1` maps atom *i* to row *i-1*, and atom 0 to row `-1`, which NumPy resolves to the *last* row. So the adjacency is the true molecular graph under the cyclic permutation σ(i) = (i−1) mod N — an isomorphic graph, but with **every atom's bonds assigned to the previous atom**.

The node features are not permuted. `mol_full_feature()` (`ligand_init.py:264`) builds them in `sorted(zip(atom_ids, atom_feats))` order, i.e. 0-based RDKit order. So features and topology disagree by one position.

This is what the model consumes. `training/dataset.py:159`:

```python
adj = mol['bond_feature'].long()
mol_edge_index = adj.nonzero(as_tuple=False).t().contiguous()
mol_edge_attr  = adj[mol_edge_index[0], mol_edge_index[1]].long()
```

and `mol_edge_index` is the `atom_edge_index` passed to every `PNAConv` in `net_v4`, `net_v5t`, `net_v7`, and now to `atom_rwse`.

### Ethanol, worked through

`CCO` — atoms 0=C, 1=C, 2=O; bonds (0,1) and (1,2). True shape is a path centred on atom 1.

```
as-is (v1-1)                    fixed (v1)
[[0 1 1]                        [[0 1 0]
 [1 0 0]                         [1 0 1]
 [1 0 0]]                        [0 1 0]]
edges: (0,1) (0,2)              edges: (0,1) (1,2)
degrees: [2 1 1]                degrees: [1 2 1]
```

The as-is graph makes **atom 0 — a terminal methyl carbon — the central node**, and atom 1 terminal.

### It contradicts itself, which makes it testable

`atom_features()` (`ligand_init.py:46`) starts with `one_of_k_encoding(atom.GetDegree(), [0..10])`. So the **first 11 dimensions of the 43-d feature vector are a one-hot of the atom's true degree** (11 + 11 + 11 + 6 + 1 + 3 = 43, matching `mol_in_channels=43`). For ethanol the feature at row 0 says degree 1 while the graph gives node 0 two neighbours.

That is a self-check you can assert without RDKit ground truth:

```python
def test_topology_matches_the_degree_feature(mol_dict):
    adj = mol_dict['bond_feature'].long()
    topo_degree = (adj != 0).sum(1)
    feat_degree = mol_dict['atom_feature'][:, :11].argmax(1)   # one_of_k_encoding(GetDegree(), 0..10)
    assert torch.equal(topo_degree, feat_degree), \
        f"{(topo_degree != feat_degree).sum().item()} atoms disagree"
```

It should fail on essentially every molecule now and pass after the fix. Worth a second test that `bond_feature` round-trips a hand-built molecule (ethanol, benzene, and something with a ring so RWSE has work to do).

### The fix

```diff
-        adj[v1 - 1, v2 - 1] = b_type
-        adj[v2 - 1, v1 - 1] = b_type
+        adj[v1, v2] = b_type
+        adj[v2, v1] = b_type
```

### Why it lands on this commit's doorstep

RWSE is a **per-node** descriptor of **topology only** — `diag((D⁻¹A)^k)` — whose entire purpose is to tell an atom which ring it sits in. Computed on this adjacency, the RWSE vector at row *i* describes atom *i+1*'s ring environment. So the new feature is guaranteed to be wrong in exactly the dimension it was added to get right, and an ablation would measure the shifted version against the baseline.

Two consequences worth being deliberate about:

- Every v4/v5t/v7 number in `METHODS_v7.md` and `figures/` was produced with this adjacency. They are real measurements of a model with a corrupted ligand graph, not of the architecture as described.
- Fixing it should *improve* results, since a bijective feature/topology misalignment destroys information the model cannot recover. This is likely free performance, not just a correctness cleanup. But it invalidates every checkpoint, so it is your call when to take it — I have deliberately not sent a PR that silently changes your inputs mid-experiment.

---

## 2. `config_v7.json` now trains the LSPE arm, so there is no baseline config left

`net_v7.py` defaults `mol_pe_mode="none"` and the branch is commented *"Baseline path — keeps the original encoder bit-identical"*. But the shipped config was flipped:

```json
"heads": 8,
"mol_pe_mode": "lspe",  "mol_pe_steps": 8, "mol_pe_dim": 16
```

So `config_v7.json` no longer reproduces v7. Anyone rerunning "v7" after 2026-08-15 gets the LSPE arm, and results from before and after are not comparable while both being called v7. `METHODS_v7.md` was not touched, so it still describes the baseline.

This matters most for the thing #1 asks for. "Add **and ablate** Positional Encoding to smile graph" needs three runs — `none`, `input`, `lspe` — and there is now exactly one config, set to the experimental arm. Suggested: keep `config_v7.json` at `"none"` and add `config_v7_pe_input.json` / `config_v7_pe_lspe.json`, or take the mode from a CLI flag so one config covers all three. Either way the ablation table wants a row per mode with the config that produced it named.

---

## 3. RWSE is recomputed every forward pass, at roughly 2500× the necessary cost

```python
def atom_rwse(edge_index, num_nodes, steps):
    A = torch.zeros(num_nodes, num_nodes, device=edge_index.device)
    ...
    for _ in range(steps - 1):
        M = M @ P
```

called from `forward()`. RWSE is a deterministic function of the molecular graph with no learnable parameters, so this is the same tensor recomputed for the same molecule on every epoch, in training and in eval.

The dense block-diagonal formulation is convenient but pays cubically for it. At the docstring's own ~1500 atoms/batch with `steps=8`:

| | FLOPs per forward |
|---|---|
| batched dense, as written: 7 × 2 × 1500³ | 4.7 × 10¹⁰ |
| per-molecule (50 mols × ~30 atoms): 50 × 7 × 2 × 30³ | 1.9 × 10⁷ |

**~2500×**, and it scales as N³ so it gets worse if the batch grows — at 4096 atoms/batch it is 1.4 TFLOP per forward with a 67 MB dense matrix. If v7 training feels slow, this is a strong candidate.

The reference implementation (`vijaydwivedi75/gnn-lspe`) precomputes positional encodings in the data pipeline for this reason. Here that means computing RWSE once per molecule in `ligand_init.py` / `dataset.py` and carrying it as a node attribute — PyG ships `torch_geometric.transforms.AddRandomWalkPE`, or keep `atom_rwse` and call it per-molecule at preprocess time. It also makes the ablation cheaper to run and removes RWSE from the training hot path entirely.

Two smaller things in the same function while you are in there:

- `A[edge_index[0], edge_index[1]] = 1.0` assumes `edge_index` carries both directions. It does here (`bond_feature` writes both triangles), but nothing asserts it, and `D⁻¹A` is only the random-walk matrix of an undirected graph. One `assert` or a `to_undirected` is cheap insurance.
- `deg = A.sum(dim=1).clamp(min=1)` gives an isolated atom an all-zero RWSE rather than return-probability 1 at every step. Only reachable for a disconnected fragment (a counterion from a `.` in a SMILES), so it depends on whether preprocessing keeps the largest fragment — worth knowing which.

---

## 4. Naming: this is LSPE without the LapEig loss, and the commit says LPSE

The code is right and the docstring is explicit — RWSE as the structural encoding, LSPE as the learnable framework, Laplacian PE deliberately rejected. The commit subject says *"via RWSE and LPSE"*, which reads as though Laplacian PE were implemented; it is not, and `pe_mode` has no such option.

More substantively, the docstring notes the auxiliary Laplacian-eigenvector loss on `p` is "not carried over ... reported as a small effect". In Dwivedi et al. (ICLR 2022) that loss is part of what makes LSPE beat PE-as-input on several benchmarks rather than a garnish. Not arguing you should add it — the honest move is to label the arm `LSPE (no LapEig loss)` in the ablation table so a reader does not compare it against the paper's LSPE numbers.

---

## Small stuff

- `net_v7.py:642`: `atom_h, mol_pe = (_enc[0], _enc[2]) if len(_enc) == 3 else (_enc[0], None)` — `mol_pe` is never read. That is consistent with the paper (`p` acts through `[h ‖ p]` inside the convs), but it is a dead local that reads as unfinished. Either drop it or add a comment saying the readout deliberately ignores `p`.
- `ligand_encoder.py` defaults `pe_steps=16`; `net_v7.py` and the config use 8. Harmless since `net_v7` always passes it, but the docstring's cost estimate is written for the local default.
- `mol_x.squeeze()` / `bond_x.squeeze()` (inherited from the v7 baseline) collapse to a 0-d tensor when there is exactly one atom or one bond in the batch. `.squeeze(-1)` is the safe form.
- Trailing whitespace on the new lines in `net_v7.py` (553-566, 640) and a tab indent in `config_v7.json`. The repo has **no tests and no CI** — the 2026-08-13 review's P1 item 5. Given #1 and #2 are both about measuring changes to this encoder, a `tests/` directory with the degree-consistency assertion above plus a shape test per `pe_mode` would pay for itself immediately, and `Huang-lab/lab-repo-template` has a CI skeleton that detects Python and runs what exists.

---

## Suggested order

1. The `bond_feature` one-liner plus the two regression tests. Nothing else means much until the graph is right.
2. Restore a baseline config, add the two PE configs.
3. Re-run the `none` / `input` / `lspe` ablation on the fixed graph — that is #1 answered with a table.
4. Move RWSE to preprocessing, which makes step 3 cheaper.

Happy to send PRs for any of these; I held off because #1 invalidates every checkpoint and that is a sequencing decision, not a code decision.

/cc @AtlasMaroc
